### PR TITLE
Add Bazel build rules to distrib packages

### PR DESCRIPTION
**Pull request by steven-johnson**
This PR builds on the work done elsewhere (e.g. github.com/halide/halide_bazel) and intends to provide a stable, official set of Bazel build rules for Halide clients. (It is *not* intended to provide a way to build Halide itself using Bazel.)

The design is such that the neecessary Bazel build rules are added to Halide distributions, such that downstream consumers can add Halide usage to their project by a few lines added to their WORKSPACE.

The build rules are modeled very closely on those used inside Google, and are expected to be stable and supported for the long-term future.

These rules should work well for Linux and OSX hosts using Bazel 0.5 or later; Windows hosts are almost-but-not-quite supported, with a handful of issues that are expected to be resolved in subsequent PRs.

(Note that the Windows distrib is currently built by a script specific to the buildbots; after this PR lands, the buildbot script will be updated separately to add the new files to the distrib.)

### CODE REVIEW

**Comment by abadams** [comment]
lgtm in theory, but I can't seem to get it to work. Let's talk in person when you're next around. We'll also need to figure out a way to do minimal testing of the demo app (maybe install bazel on the linux buildbots?).

**Comment by steven-johnson** [comment]
Since this PR, I've also learned that there is apparently a convention for surfacing Bazel build rules:

https://bazel.build/versions/master/docs/skylark/deploying.html

frankly, it looks quite a bit more complex than necessary for our purposes, and not even as convenient. Not sure how big a faux pas it would be to just ignore these guidelines.

But yeah, let's talk next week.

**Comment by steven-johnson** [comment]
PTAL

**Comment by steven-johnson** [comment]
The failure is in the performance tests (repeat_image is 2.164563 times slower than unbounded), which should be unaffected by this change.

However, it seems likely we should get at least minimal Travis testing for this in place before landing, or it will definitely get stale. I'll look into how to get Bazel testing in place for Travis.

**Comment by abadams** [comment]
I am installing bazel 0.5.2 on the linux buildbots. If you add it to "make test_apps" in such a way that it only runs when bazel exists, it'll get testing.

**Comment by steven-johnson** [comment]
Will do; I'm still experimenting with Travis. (Bazel needs Java 8 present, which apparently isn't available by default on Travis, and just asking to install the openjdk-8-jdk package fails because it isn't sure if that means you want jdk8 or jdk9...)

**Comment by steven-johnson** [comment]
Finally clean. PTAL

File: Makefile [hunk 1]
```diff
@@ -27,6 +27,8 @@ else
 endif
 endif
 
+BAZEL ?= $(shell which bazel)
```
Comment (zvookin) [inline Makefile]: Have you verified this does not cause a problem if bazel is not installed?

**Comment by zvookin** [comment]
LGTM.

File: Makefile (hunk 1 above)
Comment (steven-johnson) [inline Makefile]: Amusingly, at the moment, all of my systems have Bazel installed.

**Pull request merged**